src/hp.rs

```diff
@@ -36,15 +37,15 @@ impl Hp {
 
     pub fn heal(&mut self, value: u16) {
         if let Some(sum) = self.current.checked_add(value) {
-            self.current = sum;
+            self.current = min(sum, self.total);
         } else {
             self.current = self.total;
         }
     }
 
     pub fn damage(&mut self, value: u16) {
-        if let Some(sub) = self.current.checked_sub(value) {
-            self.current = sub;
+        if let Some(sum) = self.current.checked_sub(value) {
+            self.current = sum;
         } else {
             self.current = 0;
             self.state = HpStateEnum::Dying;
```

Comment:
The `if let` is super succinct and particularly suited for a binary outcome like this.